include/lo2s/perf/syscall/reader.hpp

```diff
@@ -68,110 +69,50 @@ class Reader : public EventReader<T>
-
-        fd_ = perf_event_open(&attr, cpu.as_scope(), -1, 0, config().cgroup_fd);
-        if (fd_ < 0)
+        TracepointEvent enter_event(0, "raw_syscalls:sys_enter");
+        TracepointEvent exit_event(0, "raw_syscalls:sys_exit");
+
+        enter_event.as_syscall();
+        exit_event.as_syscall();
+
+        try
         {
-            Log::error() << "perf_event_open for raw tracepoint failed.";
-            throw_errno();
+            enter_ev_ = enter_event.open(cpu_);
+            exit_ev_ = exit_event.open(cpu_);
         }
-
-        attr.config = tracepoint::EventFormat("raw_syscalls:sys_exit").id();
-        other_fd_ = perf_event_open(&attr, cpu.as_scope(), -1, 0, config().cgroup_fd);
-        if (other_fd_ < 0)
+        catch (const std::system_error& e)
         {
             Log::error() << "perf_event_open for raw tracepoint failed.";
             throw_errno();
-            close(fd_);
         }
 
-        try
-        {
-            if (fcntl(fd_, F_SETFL, O_NONBLOCK))
-            {
-                throw_errno();
-            }
-
-            init_mmap(fd_);
-            Log::debug() << "perf_tracepoint_reader mmap initialized";
-            if (ioctl(other_fd_, PERF_EVENT_IOC_SET_OUTPUT, fd_) == -1)
-            {
-                throw_errno();
-            }
-            if (!config().syscall_filter.empty())
-            {
-                std::vector<std::string> names;
-                std::transform(config().syscall_filter.cbegin(), config().syscall_filter.end(),
-                               std::back_inserter(names),
-                               [](const auto& elem) { return fmt::format("id == {}", elem); });
-                std::string filter = fmt::format("{}", fmt::join(names, "||"));
-
-                if (ioctl(other_fd_, PERF_EVENT_IOC_SET_FILTER, filter.c_str()) == -1)
-                {
-                    throw_errno();
-                }
-                if (ioctl(fd_, PERF_EVENT_IOC_SET_FILTER, filter.c_str()) == -1)
-                {
-                    throw_errno();
-                }
-            }
-            auto ret = ioctl(fd_, PERF_EVENT_IOC_ENABLE);
-            Log::debug() << "perf_tracepoint_reader ioctl(fd, PERF_EVENT_IOC_ENABLE) = " << ret;
-            if (ret == -1)
-            {
-                throw_errno();
-            }
-
-            ret = ioctl(other_fd_, PERF_EVENT_IOC_ENABLE);
-            Log::debug() << "perf_tracepoint_reader ioctl(fd, PERF_EVENT_IOC_ENABLE) = " << ret;
-            if (ret == -1)
-            {
-                throw_errno();
-            }
-        }
-        catch (...)
+        init_mmap(enter_ev_.get_fd());
+        Log::debug() << "perf_tracepoint_reader mmap initialized";
+
+        exit_ev_.set_output(enter_ev_);
+
+        if (!config().syscall_filter.empty())
         {
-            close(other_fd_);
-            close(fd_);
-            throw;
+            enter_ev_.set_syscall_filter();
+            exit_ev_.set_syscall_filter();
```

Comment:
Add an early exit for an empty input in set_syscall_filter() and make this call here just unconditional.